Select multiple mechanisms with path-based HttpSecurityPolicy (#52220)
quarkusio/quarkus: docs/src/main/asciidoc/security-authentication-mechanisms.adoc

[IMPORTANT]
====
Ensure that the value of the `auth-mechanism` property matches the authentication scheme supported by `HttpAuthenticationMechanism`, for example, `basic`, `bearer` or `custom`.
For example, for the `quarkus.http.auth.permission.custom.auth-mechanism=custom` expression to be effective, a custom `HttpAuthenticationMechanism` implementation must set its authentication scheme to `custom` in its `HttpAuthenticationMechanism#getCredentialTransport` method implementation.
====

==== Use HTTP Security Policy inclusive path-based authentication

When <<inclusive-authentication>> is enabled for all the HTTP paths with the `quarkus.http.auth.inclusive` configuration property, every selected authentication mechanism must provide the `SecurityIdentity`.
For example:

[source,properties]
----
quarkus.http.auth.inclusive=true

quarkus.http.auth.permission.basic-bearer.paths=/hello/basic-bearer
quarkus.http.auth.permission.basic-bearer.policy=authenticated
quarkus.http.auth.permission.basic-bearer.auth-mechanism=basic,bearer   <1>
----

> <1> The REST endpoint `/hello/basic-bearer` can be accessed by using both the xref:security-oidc-bearer-token-authentication.adoc[OIDC bearer token authentication mechanism], and the <<basic-auth>>.

Reviewer: ```suggestion
<1> The REST endpoint `/hello/basic-bearer` can be accessed only if both the xref:security-oidc-bearer-token-authentication.adoc[OIDC bearer token authentication mechanism] and the <<basic-auth>> have successfully provided `SecurityIdentity`.
```

quarkus.http.auth.permission.custom.paths=/hello/custom <3>
quarkus.http.auth.permission.custom.policy=authenticated
quarkus.http.auth.permission.custom.auth-mechanism=custom

quarkus.http.auth.permission.basic-bearer.paths=/hello/basic-bearer <4>
quarkus.http.auth.permission.basic-bearer.policy=authenticated

> quarkus.http.auth.permission.basic-bearer.auth-mechanism=basic,bearer

Author: > Thanks @michalvavrik
> 
> Before I ask more questions, can you please remind, what was the idea behind `inclusive/lax`, how does it differ from the default architecture, where, given 3 mechanisms, at least one of them must give `SecurtyIdentity` with the process stopping immediately once the identity is produced by whatever mechanism. Is the difference between the default strategy and `inclusive/lax` is that in the latter case, given for example, 3 mechanisms, if the first one produces `SecurityIdentity`, the other two will also get a chance to add to their own identity if they can ?

Yes, that was the idea. And you can see in https://quarkus.io/guides/security-authentication-mechanisms#use-inclusive-authentication-to-enable-path-based-authentication that it basically allows you to implement path-specific inclusive authentication for certain mechanisms. 

Side note: I think in light of your request to remove path-specific `inclusive` flag, we will need to keep both `lax` and that docs section even once I implement the `@HttpAuthenticationMechanism(mechanisms = { @BasicAuthentication, @MTLS })`. Because it will still allow you to say: endpoint XYZ must be authenticated by A, Z, C and if not, access must be denied.

NOTE: For consistency with various Jakarta EE specifications, it is recommended to always repeat annotations instead of relying on annotation inheritance.

> 

Author: done
